File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/InsertTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonNull;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class InsertTest {

	@Test
	void test() {
		JsonArray arr = new JsonArray();
		arr.add(1);
		arr.add(2);
		arr.add(3);
		arr.add(4);
		arr.add(5);
		arr.add(6);
		arr.add(7);
		arr.add(8);

		JsonArray res = new JsonArray();
		res.add(1);
		res.add(2);
		res.add(9);
		res.add(3);
		res.add(4);
		res.add(5);
		res.add(6);
		res.add(7);
		res.add(8);

		Insert ins = new Insert();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr, "offset", new JsonPrimitive(2), "element", new JsonPrimitive(9)))
				.setContext(Map.of()).setSteps(Map.of());

		arr = (JsonArray) ins.execute(fep).block().next().getResult().get("result");

		assertEquals(res, arr);
	}

	@Test
	void test2() {
		JsonArray arr = new JsonArray();
		arr.add(1);
		arr.add(2);
		arr.add(3);
		arr.add(4);
		arr.add(5);
		arr.add(6);
		arr.add(7);
		arr.add(8);

		JsonArray res = new JsonArray();
		res.add('a');
		res.add(1);
		res.add(2);
		res.add(3);
		res.add(4);
		res.add(5);
		res.add(6);
		res.add(7);
		res.add(8);

		Insert ins = new Insert();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr, "offset", new JsonPrimitive(0), "element", new JsonPrimitive('a')))
				.setContext(Map.of()).setSteps(Map.of());

		arr = (JsonArray) ins.execute(fep).block().next().getResult().get("result");

		assertEquals(res, arr);

	}

	@Test
	void test3() {
		JsonArray arr = new JsonArray();

		JsonArray res = new JsonArray();
		res.add('a');

		Insert ins = new Insert();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr, "offset", new JsonPrimitive(0), "element", new JsonPrimitive('a')))
				.setContext(Map.of()).setSteps(Map.of());

		arr = (JsonArray) ins.execute(fep).block().next().getResult().get("result");

		assertEquals(res, arr);

	}

	@Test
	void test4() {

		JsonArray res = new JsonArray();
		res.add('a');

		Insert ins = new Insert();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository()).setArguments(
						Map.of("source", JsonNull.INSTANCE, "offset", new JsonPrimitive(0), "element",
								new JsonPrimitive('a')))
				.setContext(Map.of()).setSteps(Map.of());

		StepVerifier.create(ins.execute(fep)).expectError(KIRuntimeException.class).verify();

	}

}

File: kirun-js/__tests__/engine/function/system/array/InsertTest.ts
import { Insert } from '../../../../../src/engine/function/system/array/Insert';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';
import { KIRunFunctionRepository, KIRunSchemaRepository } from '../../../../../src';

test('Insert of Test 1', async () => {
    let ins: Insert = new Insert();

    let array: any[] = [];

    array.push('test');
    array.push('Driven');
    array.push('developement');
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('I');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Insert.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
                [Insert.PARAMETER_INT_OFFSET.getParameterName(), 4],
                [Insert.PARAMETER_ANY.getParameterName(), ['this is an array']],
            ]),
        )
        .setSteps(new Map([]))
        .setContext(new Map([]));

    let res: any[] = [];

    res.push('test');
    res.push('Driven');
    res.push('developement');
    res.push('I');
    res.push(['this is an array']);
    res.push('am');
    res.push('using');
    res.push('eclipse');
    res.push('I');
    res.push('to');
    res.push('test');
    res.push('the');
    res.push('changes');
    res.push('with');
    res.push('test');
    res.push('Driven');
    res.push('developement');

    expect((await ins.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(res);
});

test('Insert of Test 2', async () => {
    let ins: Insert = new Insert();

    let arr: any[] = [];

    arr.push(1);
    arr.push(2);
    arr.push(3);
    arr.push(4);
    arr.push(5);
    arr.push(6);
    arr.push(7);
    arr.push(8);

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Insert.PARAMETER_ARRAY_SOURCE.getParameterName(), arr],
                [Insert.PARAMETER_INT_OFFSET.getParameterName(), 2],
                [Insert.PARAMETER_ANY.getParameterName(), ['this is an array']],
            ]),
        )
        .setSteps(new Map([]))
        .setContext(new Map([]));

    let res: any[] = [];

    res.push(1);
    res.push(2);
    res.push(['this is an array']);
    res.push(3);
    res.push(4);
    res.push(5);
    res.push(6);
    res.push(7);
    res.push(8);
    expect((await ins.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(res);
});

test('Insert of Test 3', async () => {
    let ins: Insert = new Insert();

    let arr: any[] = [];

    arr.push(1);
    arr.push(2);
    arr.push(3);
    arr.push(4);
    arr.push(5);
    arr.push(6);
    arr.push(7);
    arr.push(8);

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Insert.PARAMETER_ARRAY_SOURCE.getParameterName(), arr],
                [Insert.PARAMETER_INT_OFFSET.getParameterName(), 0],
                [Insert.PARAMETER_ANY.getParameterName(), ['this is an array']],
            ]),
        )
        .setSteps(new Map([]))
        .setContext(new Map([]));

    let res: any[] = [];

    res.push(['this is an array']);
    res.push(1);
    res.push(2);
    res.push(3);
    res.push(4);
    res.push(5);
    res.push(6);
    res.push(7);
    res.push(8);
    expect((await ins.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(res);
});

test('Insert of Test 4', async () => {
    let ins: Insert = new Insert();

    let arr: any[] = [];

    arr.push(1);
    arr.push(2);
    arr.push(3);
    arr.push(4);
    arr.push(5);
    arr.push(6);
    arr.push(7);
    arr.push(8);

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Insert.PARAMETER_ARRAY_SOURCE.getParameterName(), arr],
                [Insert.PARAMETER_INT_OFFSET.getParameterName(), arr.length],
                [Insert.PARAMETER_ANY.getParameterName(), ['this is an array']],
            ]),
        )
        .setSteps(new Map([]))
        .setContext(new Map([]));

    let res: any[] = [];

    res.push(1);
    res.push(2);
    res.push(3);
    res.push(4);
    res.push(5);
    res.push(6);
    res.push(7);
    res.push(8);
    res.push(['this is an array']);

    expect((await ins.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(res);
});

test('Insert of Test 5', async () => {
    let ins: Insert = new Insert();

    let arr: any[] = [];

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Insert.PARAMETER_ARRAY_SOURCE.getParameterName(), arr],
                [Insert.PARAMETER_ANY.getParameterName(), ['this is an array']],
            ]),
        )
        .setSteps(new Map([]))
        .setContext(new Map([]));

    let res: any[] = [];

    res.push(['this is an array']);

    expect((await ins.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(res);
});

test('Insert of Test 6', async () => {
    let ins: Insert = new Insert();

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Insert.PARAMETER_ARRAY_SOURCE.getParameterName(), null],
                [Insert.PARAMETER_INT_OFFSET.getParameterName(), 0],
                [Insert.PARAMETER_ANY.getParameterName(), ['this is an array']],
            ]),
        )
        .setSteps(new Map([]))
        .setContext(new Map([]));

    let res: any[] = [];

    res.push(['this is an array']);

    await expect(ins.execute(fep)).rejects.toThrow();
});
